### PR TITLE
fix nullable Part I

### PR DESCRIPTION
**Pull request by Jim8y**
fix https://github.com/neo-project/neo-devpack-dotnet/pull/1041#pullrequestreview-2065152303

### CODE REVIEW

File: tests/Neo.Compiler.CSharp.UnitTests/TestingArtifacts/Contract_NULL.cs [hunk 1]
```diff
@@ -22,25 +22,25 @@ public abstract class Contract_NULL : Neo.SmartContract.Testing.SmartContract
     /// Unsafe method
     /// </summary>
     [DisplayName("equalNotNullA")]
-    public abstract bool? EqualNotNullA(byte[]? value);
+    public abstract bool? EqualNotNullA(IList<object>? value);
```
Comment (Jim8y) [inline tests/Neo.Compiler.CSharp.UnitTests/TestingArtifacts/Contract_NULL.cs:25]: This is automatically generated by the test engine, not me.

File: src/Neo.SmartContract.Testing/Native/ContractManagement.cs
Comment (cschuchardt88) [inline src/Neo.SmartContract.Testing/Native/ContractManagement.cs:1]: Can you fix `The event 'ContractManagement.OnDeploy' is never used` in this file too?

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_Array.cs [hunk 2]
```diff
@@ -99,7 +101,7 @@ public static object[] TestEmptyArray()
             return sarray;
         }
 
-        public static object TestStructArrayInit()
+        public static object? TestStructArrayInit()
```
Comment (cschuchardt88) [inline tests/Neo.Compiler.CSharp.TestContracts/Contract_Array.cs:104]: May want to check `Unreachable code detected` in this method.

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_StaticByteArray.cs [hunk 3]
```diff
@@ -6,7 +6,9 @@ namespace Neo.Compiler.CSharp.TestContracts
     public class Contract_StaticByteArray : SmartContract.Framework.SmartContract
     {
         [DisplayName("TestEvent")]
-        public static event Action<byte[]> OnEvent;
+#pragma warning disable CS0067 // Event is never used
+        public static event Action<byte[]> OnEvent = default!;
```
Comment (cschuchardt88) [inline tests/Neo.Compiler.CSharp.TestContracts/Contract_StaticByteArray.cs:10]: Fix `The field 'Contract_StaticByteArray.OnEvent' is assigned but its value is never used`?

File: src/Neo.SmartContract.Testing/Native/ContractManagement.cs [hunk 4]
```diff
@@ -24,12 +24,16 @@ public abstract class ContractManagement : SmartContract
     public delegate void delDestroy(UInt160 Hash);
 
     [DisplayName("Destroy")]
+#pragma warning disable CS0067 // Event is never used
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/Native/ContractManagement.cs:27]: This event can be used

File: src/Neo.SmartContract.Testing/Native/GAS.cs [hunk 5]
```diff
@@ -14,10 +14,10 @@ public abstract class GAS : SmartContract, TestingStandards.INep17Standard
     #endregion
 
     #region Events
-
+#pragma warning disable CS0067 // Event is never used
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/Native/GAS.cs:17]: This event can be used

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]
## Missed some

### tests/Neo.Compiler.CSharp.TestContracts/Contract_ABISafe.cs
The field 'Contract_ABISafe.s' is assigned but its value is never used

### src/Neo.SmartContract.Analyzer/CollectionTypesUsageAnalyzer.cs
Dereference of a possibly null reference.

### src/Neo.SmartContract.Analyzer/BigIntegerUsingUsageAnalyzer.cs
Possible null reference argument for parameter 'root' in 'Document Document.WithSyntaxRoot(SyntaxNode root)'.

File: src/Neo.SmartContract.Analyzer/NepStandardAnalyzer.cs [hunk 6]
```diff
@@ -107,7 +107,9 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
             var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
             var diagnostic = context.Diagnostics.First();
             var diagnosticSpan = diagnostic.Location.SourceSpan;
+#pragma warning disable CS8602 // Dereference of a possibly null reference.
             var attributeSyntax = root.FindNode(diagnosticSpan).AncestorsAndSelf().OfType<AttributeSyntax>().FirstOrDefault();
+#pragma warning restore CS8602 // Dereference of a possibly null reference.
```
Comment (shargon) [inline src/Neo.SmartContract.Analyzer/NepStandardAnalyzer.cs:112]:
```suggestion
            var attributeSyntax = root?.FindNode(diagnosticSpan).AncestorsAndSelf().OfType<AttributeSyntax>().FirstOrDefault();
```

File: src/Neo.SmartContract.Testing/Native/ContractManagement.cs (hunk 4 above)
Comment (Jim8y) [reply src/Neo.SmartContract.Testing/Native/ContractManagement.cs:27]: This just disable the warning, wont affect its function.

**Review by shargon — APPROVED** [review]

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_Array.cs (hunk 2 above)
Comment (Jim8y) [reply tests/Neo.Compiler.CSharp.TestContracts/Contract_Array.cs:104]: That code is weird,,,,,kinda like for loop is not necessary at all,,,,,,

**Comment by Jim8y** [comment]
@shargon  please allow me to stop this one from here, too many warnings (200 to go) to fix, will do the remaining in another pr.

File: src/Neo.Compiler.CSharp/MethodConvert/Expression/Expression.cs [hunk 7]
```diff
@@ -55,24 +55,24 @@ private void ConvertExpression(SemanticModel model, ExpressionSyntax syntax, Syn
                 {
                     //complex types like UInt160 at compile time to avoid runtime overhead.
                     case "Neo.SmartContract.Framework.UInt160":
-                        var strValue = (string)value;
+                        var strValue = (string)value!;
```
Comment (vncoelho) [inline src/Neo.Compiler.CSharp/MethodConvert/Expression/Expression.cs:58]: where do we ensure value is never null?

Comment (Jim8y) [reply src/Neo.Compiler.CSharp/MethodConvert/Expression/Expression.cs:58]: we dont, this just fix the warning. the code still run as past.

Comment (Jim8y) [reply src/Neo.Compiler.CSharp/MethodConvert/Expression/Expression.cs:58]: > where do we ensure value is never null?

i used to do null check, but turns out it just work as !

**Review by cschuchardt88 — APPROVED** [review]

**Pull request merged**